src/java/org/apache/cassandra/io/sstable/format/bti/PartitionIterator.java

```diff
@@ -0,0 +1,246 @@
+            if (partitionIterator.nextKey != null && !(partitionIterator.nextKey.compareTo(left) > inclusiveLeft))
+            {
+                partitionIterator.readNext();
+            }
+            partitionIterator.advance();
+            return partitionIterator;
+        }
+        catch (IOException | RuntimeException ex)
+        {
+            if (partitionIterator != null)
+            {
+                partitionIterator.close();
+            }
+            else
+            {
+                Throwables.closeNonNullAndAddSuppressed(ex, rowIndexFileCopy, dataFileCopy, partitionIndexCopy);
+            }
+            throw ex;
+        }
+    }
+
+    static PartitionIterator create(PartitionIndex partitionIndex, IPartitioner partitioner, FileHandle rowIndexFile, FileHandle dataFile) throws IOException
+    {
+        return create(partitionIndex, partitioner, rowIndexFile, dataFile, partitionIndex.firstKey(), -1, partitionIndex.lastKey(), 0);
+    }
+
+    static PartitionIterator empty(PartitionIndex partitionIndex)
+    {
+        return new PartitionIterator(partitionIndex.sharedCopy());
+    }
+
+    private PartitionIterator(PartitionIndex partitionIndex, IPartitioner partitioner, FileHandle rowIndexFile, FileHandle dataFile,
+                              PartitionPosition left, PartitionPosition right, int exclusiveRight)
+    {
+        super(partitionIndex, left, right);
+        this.partitionIndex = partitionIndex;
+        this.partitioner = partitioner;
+        this.limit = right;
+        this.exclusiveLimit = exclusiveRight;
+        this.rowIndexFile = rowIndexFile;
+        this.dataFile = dataFile;
+    }
+
+    private PartitionIterator(PartitionIndex partitionIndex)
+    {
+        super(partitionIndex, partitionIndex.firstKey(), partitionIndex.firstKey());
+        this.partitionIndex = partitionIndex;
+        this.partitioner = null;
+        this.limit = partitionIndex.firstKey();
+        this.exclusiveLimit = -1;
+        this.rowIndexFile = null;
+        this.dataFile = null;
+
+        this.currentEntry = null;
+        this.currentKey = null;
+        this.nextEntry = null;
+        this.nextKey = null;
+    }
+
+    @Override
+    public void close()
+    {
+        Throwable accum = null;
+        accum = Throwables.close(accum, immutableListWithFilteredNulls(partitionIndex, dataFile, rowIndexFile));
+        accum = Throwables.close(accum, immutableListWithFilteredNulls(dataInput, indexInput));
+        accum = Throwables.perform(accum, super::close);
+        Throwables.maybeFail(accum);
+    }
+
+    public DecoratedKey decoratedKey()
+    {
+        return currentKey;
+    }
+
+    public ByteBuffer key()
+    {
+        return currentKey.getKey();
+    }
+
+    @Override
+    public long dataPosition()
+    {
+        return currentEntry != null ? currentEntry.position : -1;
+    }
+
+    @Override
+    public long keyPositionForSecondaryIndex()
```

Comment:
Aside: SAI isn't even using this anymore, so I guess it'll be nice when we can remove SASI from the codebase...